design/Implemented/existing-resource-policy_design.md

```diff
@@ -55,10 +55,11 @@ skip restoration.
    - Changed resources: Velero will first try to patch the changed resource, Now if the patch:
        - succeeds: Then the in-cluster resource gets updated with the labels as well as the resource diff
        - fails: Velero adds a restore warning and tries to just update the backup/restore labels on the resource, if the labels patch also fails then we add restore error.
-3. `recreate`:  If resource already exists, then Velero will delete it and recreate the resource.
+3. `recreate`:  Similar to `update` but if resource already exists and is immutable (such as Pods), then Velero attempt to recreate the resource by deleting it first.
```

Comment:
@anshulahuja98 without any special code for specific resource types (which I don't think we have right now), we would only delete and recreate something if:
1) the kubernetes metadata in-cluster is different from what's in-cluster
2) the attempt to patch the resource failed.

Also, at the point where the delete-and-recreate happens, all plugins, etc. have already run. However, it's possible that this will actually work *if* there are immutable fields that we're attempting to patch here. Otherwise, there may be value in making an exception for the PVC itself in the code. I think that may be better handled as a follow-on change to handle the specific PVC use case, since there may be a number of edge cases to deal with -- CSI-with-datamover, CSI-without-datamover, native snapshots, etc.

Reply:
I thought further on this. I am now trying to think from the angle of, what is recreate exactly solving  for the end user. I have a worry that it might lead to the cluster in a bad shape than expected restore with the current design approach of deleting if patch fails.

For example we have a `deployment` and a `pvc` attached to it. As per the RestoreORder used by velero today it will first go to PVC -> patch will fail most likely(let's assume it will) -> velero will try to delete PVC but it will never complete since PVC is mounted on deployment -> velero will now move forward after just marking PVC for deletion -> Velero will now try to restore deployment (where let's assume recreate kicks in ) -> Deployment gets deleted -> leading to the `terminating state PVC `getting cleaned up -> Deployment created but PVC is gone / or if recreate didn't kick in for deployment, PVC will be in terminating state.

As a result we caused the user data loss by marking the current PVC deleted while not restoring the data back. The cluster would be in a bad state. IMO: this is kind of a P0 in my opinion to avoid users shooting themselves in the foot. 

I believe from k8s core API this is very basic use case which we should think how it can be supported as part of the design or even if a follow on PR can solve this properly if we special case PVC. The question is - is the design extensible in current velero to support it properly?

Reply:
I have a similar concern with problems because of complex relationships between objects:

- An object may belong to an application which consists of multiple objects
- If one object exists at the time of restore, it may indicate that the entire application is running with many other objects
- We don't know the relationship of these objects, there may be one possibility -- one other object is monitoring the kind of the current object, if it is missing, a new one with the same kind & name will be created
- Then even though we successfully delete the object as part of the recreate logic, there may still be further problems:
  - Either the creation of the object as the other part of the recreate logic will fail because the application recreates the object first
  - Or the existing application itself goes to a chaos or modifies the newly created object erroneously which cause the object unusable
- Finally, it looks like we have successfully created all the objects due for the restore, but the restored application cannot work appropriately

Totally speaking, this indicates that it is more safe to delete the application/namespace as a whole instead of deleting single objects one by one for this recreate logic, because the latter is not in a consistent manner.

Reply:
@anshulahuja98 "I thought further on this. I am now trying to think from the angle of, what is recreate exactly solving for the end user. I have a worry that it might lead to the cluster in a bad shape than expected restore with the current design approach of deleting if patch fails."

The main thing this is supposed to resolve is if the backup contains different metadata for immutable fields (or for immutable objects). If the patch attempts to modify an immutable field or an immutable resource, it will fail.

As for the concern that this will break things in some cases (PVCs, etc.), I had similar concerns, which is why my original recommendation was to implement this after we had implemented the per-resource-type overrides from the original design document.  Right now, policy is all-or-none. We either do nothing across the board, or we attempt to patch resources across the board, or (with this PR), attempt to delete-and-recreate for everything that fails patching. I'm starting to think we may want to first implement the per-resource override functionality and then the recreate policy. That would allow us, for example, to set the update/patch policy for (almost) everything, but then recreate for just the specific resource type that the user has hit patch failures with. This will limit the potential breakage from the recreate policy.

Reply:
> we may want to first implement the per-resource override functionality and then the recreate policy.

That's an option

Reply:
Can you perhaps add more details about the per-resource-type override and what that means exactly? I can't seem to find the design doc mentioned here.

Also additionally, +1 to what @Lyndon-Li  mentioned in another comment below. 
I'll also prefer a Namespace delete & create than a per resource type.
where we can perhaps bulk delete namespaces and poll for them in parallel for 5mins or so and then proceed to restore.

For cluster resources on the other hand, there might be more edge cases. But namespace deletion + current recreate for cluster scope might be a better solution. 


Given the issues we are facing with CSI not snapshotting in parallel leading to longer backup times, deletion related polling should also be more efficient than what it is now. ( current impl has 2 min wait per resource, if let's say 50 resources are stuck for deletion due to some reason, restore time shoots up to 100mins. We need either some parallel effort or top level deletion(namespace).

Reply:
@anshulahuja98 here's the section in the design doc I was referring to: https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/existing-resource-policy_design.md#approach-3-combination-of-approach-1-and-approach-2

Reply:
went through it, I am not sure if this will still solve the ordering issue.
what are your thoughts on namespace override.

Or perhaps a different approach all together. (thinking out loud)
We can perhaps iterate through all resources in restore flow -> We'll just Delete whatever resources are in the "backup" and in the cluster.
Then run restore again with none as the resourcepolicy.

It might be a proper recreate.

Reply:
this might work for purely k8s objects backup.. if there are other CRDs for other controllers, the order of creation of objects may matter and therefore the chances that everything work is not guaranteed.. so perhaps we introduce it as an alpha feature?

Reply:
your point is valid for other controllers as well.
But a mass deletion before creation of resources / deleting entire namespaces is still more likely to succeed. Since atleast resources do  not stay stuck in deletion for the most part.
I believe we need to go back to design phase for this. Introducing current impl as alpha feature is not useful for any real customer scenario and will introduce more complications in understanding. 
But that is just my opinion on this particular feature. Others please share your thoughts.

Reply:
It might be useful to check compete offerings on how they tackle this issue to come up with the design.

Reply:
unless we also mass remove finalizers